blockprotocol/blockprotocol
# Add helper function for fully recomputing edges in Subgraph

## 🌟 What is the purpose of this PR?



This PR changes how we update entities to take into account that edges may need to recompute. It does so by removing all edges and recompute them through the `infer` functions.

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203623179643290/1204145497706392/f) _(internal)_

## 🚫 Blocked by




## 🔍 What does this change?




- Added new `inferSubgraphEdges` helper to recompute all edges of the graph given the vertices
- Delete edges on update
- Recompute the edges on update through `inferSubgraphEdges`

## ❓ How to test this?



I suggest the following test setup:
- checkout `asa/temp-updateentity-mbd` at commit `5c8668d98716c0431939df090ddb323c7a3dedbd`
- Build `@blockprotocol/graph` and `mock-block-dock`
- `yarn dev` from the block at `blocks/aio-demo` 

Operations on the subgraph through the dev block should result in the right edges being present.

## 📹 Demo


This demo may be hard to follow without a description.
What happens here is I've seeded entity types into the demo and added 1 person, 1 org and a link between them.
I show the state of the edges where the `Person` has `IS_OF_TYPE` edges pointing to the `Person` entity type in the beginning. I am going to change the `Person` to have the `Thing` entity type instead. When this update happens, the edges are fully recomputed. The, previously, `Person`  entity is now a `Thing` with its edges adjusted.

https://user-images.githubusercontent.com/6988668/224391588-09598f52-2e9f-4ac7-9d6a-0899f76fe562.mp4
